src/main/java/org/opensearch/knn/index/memory/NativeMemoryCacheManager.java

```diff
@@ -0,0 +1,298 @@
+    private long maxWeight;
+
+    NativeMemoryCacheManager() {
+        this.executor = Executors.newSingleThreadExecutor();
+        this.cacheCapacityReached = new AtomicBoolean(false);
+        this.maxWeight = Long.MAX_VALUE;
+        initialize();
+    }
+
+    /**
+     * Make sure we just have one instance of cache.
+     *
+     * @return NativeMemoryCacheManager instance
+     */
+    public static synchronized NativeMemoryCacheManager getInstance() {
+        if (INSTANCE == null) {
+            INSTANCE = new NativeMemoryCacheManager();
+        }
+        return INSTANCE;
+    }
+
+    private void initialize() {
+        CacheBuilder<String, NativeMemoryAllocation> cacheBuilder = CacheBuilder.newBuilder()
+                .recordStats()
+                .concurrencyLevel(1)
+                .removalListener(this::onRemoval);
+
+        if(KNNSettings.state().getSettingValue(KNNSettings.KNN_MEMORY_CIRCUIT_BREAKER_ENABLED)) {
+            maxWeight = KNNSettings.getCircuitBreakerLimit().getKb();
+            cacheBuilder.maximumWeight(maxWeight).weigher((k, v) -> (int)v.getSizeInKb());
+        }
+
+        if(KNNSettings.state().getSettingValue(KNNSettings.KNN_CACHE_ITEM_EXPIRY_ENABLED)) {
+            long expiryTime = ((TimeValue) KNNSettings.state()
+                    .getSettingValue(KNNSettings.KNN_CACHE_ITEM_EXPIRY_TIME_MINUTES)).getMinutes();
+            cacheBuilder.expireAfterAccess(expiryTime, TimeUnit.MINUTES);
+        }
+
+        cacheCapacityReached = new AtomicBoolean(false);
+
+        cache = cacheBuilder.build();
+    }
+
+    /**
+     * Evicts all entries from the cache and rebuilds.
+     */
+    public synchronized void rebuildCache() {
+        logger.info("KNN Cache rebuilding.");
+        executor.execute(() -> {
+            cache.invalidateAll();
+            initialize(); }
+        );
+    }
+
+    @Override
+    public void close() {
+        executor.shutdown();
+    }
+
+    /**
+     * Getter for current cache size in Kilobytes.
+     *
+     * @return current size of the cache
+     */
+    public long getCacheSizeInKilobytes() {
+        return cache.asMap().values().stream().mapToLong(NativeMemoryAllocation::getSizeInKb).sum();
+    }
+
+    /**
+     * Returns the current size of an index in the cache in KiloBytes.
+     *
+     * @param indexName Name if index to get the weight for
+     * @return Size of the index in the cache in kilobytes
+     */
+    public Long getIndexSizeInKilobytes(final String indexName) {
+        Validate.notNull(indexName, "Index name cannot be null");
+        return cache.asMap().values().stream()
+                .filter(nativeMemoryAllocation -> nativeMemoryAllocation instanceof NativeMemoryAllocation.IndexAllocation)
+                .filter(indexAllocation -> indexName.equals(((NativeMemoryAllocation.IndexAllocation) indexAllocation).getOpenSearchIndexName()))
+                .mapToLong(NativeMemoryAllocation::getSizeInKb)
```

Comment:
Will there be more than one entry for same OpenSearchIndexName?

Reply:
Yes --> there is one (or more) entries per segment, multiple segments per shard, multiple shards per OpenSearch index. So in most cases, there will be more than one entry.